r1cs/gadgets/std/src/bits/macros.rs

```diff
@@ -0,0 +1,2532 @@
+                    let mut multi_eq = MultiEq::new(&mut cs);
+                    let _ = self.sub_noborrow(&mut multi_eq, other)?;
+
+                    Ok(())
+                }
+
+                // Return little endian representation of self. Will be removed when to_bits_le and
+                // from_bits_le will be merged.
+                pub fn into_bits_le(&self) -> Vec<Boolean> {
+                    self.bits.to_vec()
+                }
+
+                // Construct self from its little endian bit representation. Will be removed when
+                // to_bits_le and from_bits_le will be merged.
+                pub fn from_bits_le<ConstraintF, CS>(cs: CS, bits: &[Boolean]) -> Result<Self, SynthesisError>
+                where
+                    ConstraintF: PrimeField,
+                    CS: ConstraintSystemAbstract<ConstraintF>,
+                {
+                    let be_bits = bits.iter().rev().map(|el| *el).collect::<Vec<_>>();
+                    Self::from_bits(cs, &be_bits)
+                }
+
+                // Construct Self from a little endian byte representation, provided in the form of
+                // byte gadgets
+                pub fn from_bytes(bytes: &[UInt8]) -> Result<Self, SynthesisError> {
+                    assert!(bytes.len()*8 <= $bit_size);
+                    let mut bits = Vec::with_capacity($bit_size);
+                    let mut value: Option<$native_type> = Some(0);
+                    for (i, byte) in bytes.iter().enumerate() {
+                        value = match byte.get_value() {
+                            Some(val) => value.as_mut().map(|v| {
+                                let val_native_type: $native_type = val.into();
+                                *v |= val_native_type << (i*8);
+                                *v}),
+                            None => None,
+                        };
+                        bits.append(&mut byte.into_bits_le());
+                    }
+
+                    // pad with 0 bits to get to $bit_size
+                    while bits.len() != $bit_size {
+                        bits.push(Boolean::constant(false));
+                    }
+
+                    Ok(Self{
+                        bits,
+                        value,
+                    })
+                }
+
+                /// This function allows to multiply `self` and `other` with a variant of
+                /// double & add algorithm.
+                /// It is useful when the field ConstraintF is too small to employ the much more
+                /// efficient algorithm employed in multiplication functions of the UIntGadget.
+                /// If `no_carry` is true, then the function checks that the multiplication
+                /// does not overflow, otherwise modular multiplication with no overflow checking
+                /// is performed
+                fn mul_with_double_and_add<ConstraintF, CS>(&self, mut cs: CS, other: &Self,
```

Comment:
It seems to me very expensive and most likely we will never use it (it happens only when multiplying 2 UInt128 when using 255 bits fields). I would even suggest to remove it and remove even the UInt128 macro impl at this point (we will never use it most likely)

Reply:
We decided to keep the `UInt128` type for the moment